boofuzz/socket_connection.py

```diff
@@ -98,7 +99,7 @@ def __init__(
         self._recv_timeout = recv_timeout
         self._send_timeout = send_timeout
         self.proto = proto.lower()
-        self.ethernet_proto = ethernet_proto
+        self.ethernet_proto = ETH_P_IP if ethernet_proto is None and proto == "raw-l3" else ethernet_proto
```

Comment:
How about this instead?
```python
if ethernet_proto is None:
    if proto == "raw-l2":
        self.ethernet_proto = 0
    elif proto == "raw-l3":
        self.ethernet_proto = ETH_P_IP
else:
    self.ethernet_proto = ethernet_proto
```
Just a code style question but I think we could gain some readability here.

Reply:
I'll change that, no problem. However, in the grand scheme, it may be a good idea to split the whole Target class and get rid of such argument conflicts and if-if-else-else ladders entirely.

Keep in mind though that this would introduce some breaking changes [0], as it would also mean that the `proto` argument is then surplus - you're directly instanciating a, say, `UdpTarget` instead of `Target(proto=udp` and so on. Having that cleanly OOPed would also make the adding (and testing) of new targets easier as it'd spare me to edit the `Target` class if I wanted to add support for Token Ring or CAN or whatever; and implementing a new target class would probably have less possibilities for side effects than now.

[0] this could always be kinda remedied by naming the new base something like `BaseTarget` and having `Target` as a shim to select the correct implementation while printing deprecated warnings.

Reply:
Good point about splitting the Target class, this would indeed make things easier. Although, I'm not sure if the Target class is the right one to split as it also contains other arguments than just the connection.
Wouldn't it be better to split and rename the SocketConnection class instead so that we can do something like this?
```python
sess = Session(
    target=Target(
        connection=TCPConnection(host=127.0.0.1, port=1234),
        netmon=...),

sess = Session(
    target=Target(
        connection=RawL2Connection(interface="lo", ethernet_proto='0x0800'),
        netmon=...),
```
Either way, we'd have to find a way to keep it backwards compatible.

Reply:
Ah, sure, I've meant the SocketConnection class. That's what you get when you run low on sleep.

But yeah, we could totally do that and keep the SocketConnection constructor as-is as kind of a factory for the splitted classes, thus deferring the break until sufficient time has passed. Seeing as the SocketConnection currently is only 350-odd lines, I'll give that a go on Monday.